src/services/pcn-packetcapture/src/Packetcapture.cpp

```diff
@@ -0,0 +1,358 @@
+typedef u_int bpf_u_int32;
+
+struct pcap_file_header {
+  bpf_u_int32 magic;      /* magic number */
+  u_short version_major;  /* major version number */
+  u_short version_minor;  /* minor version number */
+  bpf_int32 thiszone;     /* GMT to local correction */
+  bpf_u_int32 sigfigs;    /* accuracy of timestamps */
+  bpf_u_int32 snaplen;    /* max length of captured packets, in octets */
+  bpf_u_int32 linktype;   /* data link type */
+};
+
+struct pcap_pkthdr {
+  bpf_u_int32 ts_sec;
+  bpf_u_int32 ts_usec;  
+  bpf_u_int32 caplen;     /* number of octets of packet saved in file */
+  bpf_u_int32 len;        /* actual length of packet */
+};
+
+Packetcapture::Packetcapture(const std::string name, const PacketcaptureJsonObject &conf)
+  : TransparentCube(conf.getBase(), { packetcapture_code_ingress }, { packetcapture_code_egress }),
+    PacketcaptureBase(name) {
+  logger()->info("Creating Packetcapture instance");
+    setCapture(conf.getCapture());
+    setAnomimize(conf.getAnomimize());
+
+  if (conf.dumpIsSet()) {
+    setDump(conf.getDump());
+  }
+
+  setNetworkmode(conf.getNetworkmode());
+  addFilters(conf.getFilters());
+  addGlobalheader(conf.getGlobalheader());
+  CapStatus = (uint8_t) conf.getCapture();
+
+  auto t_filters_in = get_array_table<filters_table>("filters_map", 0, ProgramType::INGRESS);
+  auto t_filters_out = get_array_table<filters_table>("filters_map", 0, ProgramType::EGRESS);
+  filters_table ft_init;
+  ft_init.dst_port_flag = ft_init.src_port_flag = ft_init.network_filter_src_flag = ft_init.network_filter_dst_flag = ft_init.l4proto_flag = false;
+  t_filters_in.set(0x0, ft_init);
+  t_filters_out.set(0x0, ft_init);
+  writeHeader = true;
+  dt = "";
+}
+
+void Packetcapture::writeDump(const std::vector<uint8_t> &packet){
+  PacketJsonObject pj;
+  auto p = std::shared_ptr<Packet>(new Packet(*this, pj));
+  struct timeval tp;
+  std::streamsize len;
+
+  if( dt == "" ){
+    time_t now = time(0);
+    tm *ltm = localtime(&now);
+    std::stringstream tmstmp;
+    tmstmp << ltm->tm_mday << "-" << ltm->tm_mon << "-" << (ltm->tm_year + 1900) << "-" << ltm->tm_hour << ":" << ltm->tm_min << ":" << ltm->tm_sec;
+    dt = tmstmp.str();
+  }
+  
+  myFile.open("capture_"+ dt +".pcap", std::ios::binary | std::ios::app);
+
+  if( writeHeader == true ){
+    struct pcap_file_header *pcap_header = new struct pcap_file_header;
+    pcap_header->magic = global_header->getMagic();
+    pcap_header->version_major = global_header->getVersionMajor();
+    pcap_header->version_minor = global_header->getVersionMinor();
+    pcap_header->thiszone = global_header->getThiszone();   //timestamp are always in GMT
+    pcap_header->sigfigs = global_header->getSigfigs();
+    pcap_header->snaplen = filters->getSnaplen();
+    pcap_header->linktype = global_header->getLinktype();
+
+    myFile.write(reinterpret_cast<const char*>(pcap_header), sizeof(*pcap_header));
+    writeHeader = false;
+  }
+  gettimeofday(&tp, NULL);
```

Comment:
If I understand correctly, we assign the timestamp to the packet in the control plane. This may not be correct, especially if we want to have a precise timestamping (e.g., in case of forensic investigations).
We should timestamp the packet while in dataplane, in order to be more precise, and send the packet in userspace including this number.